src/context/values.js

```diff
@@ -0,0 +1,720 @@
+    const usedByKey = this.usedByKey_;
+    if (!usedByKey) {
+      return;
+    }
+    if (this.isConnected_()) {
+      // Ping all properties for recalculation when the tree is connected.
+      usedByKey.forEach((used) => {
+        const {prop} = used;
+        this.ping(prop, true);
+      });
+    } else {
+      // On disconnect, only do minimal possible work: disconnect recursive
+      // subscribers to ensure that they are not leaked.
+      usedByKey.forEach((used) => {
+        const {prop} = used;
+        if (isRecursive(prop)) {
+          this.updateParentValueNode_(used, null);
+        }
+      });
+    }
+  }
+
+  /**
+   * Used for `deepScan` scanner to notify the subtree that the specified
+   * property has changed and needs to be recalculated.
+   *
+   * Scans are relatively common and this method exists (as opposed to be
+   * inlined) only to avoid frequent function allocation.
+   *
+   * @param {!ContextProp} prop
+   * @return {boolean}
+   * @protected Necessary for cross-binary access.
+   */
+  scan(prop) {
+    this.ping(prop, true);
+    if (!isRecursive(prop)) {
+      // Stop the deepscan. The prop doesn't propagate.
+      return false;
+    }
+    if (this.has(prop)) {
+      // Stop the deepscan. The node will propagate changes downstream.
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Used for `deepScan` scanner to notify the subtree to recalculate all
+   * properties.
+   *
+   * Scans are relatively common and this method exists (as opposed to be
+   * inlined) only to avoid frequent function allocation.
+   *
+   * @param {!Array<string>} scheduled The already scheduled props.
+   * @return {!Array<string>} The new scheduled props.
+   * @protected
+   */
+  scanAll(scheduled) {
+    let newScheduled = null;
+    const usedByKey = this.usedByKey_;
+    if (usedByKey) {
+      usedByKey.forEach((used) => {
+        const {prop} = used;
+        // Only ping unhandled props.
+        if ((newScheduled || scheduled).indexOf(prop.key) == -1) {
+          this.ping(prop, true);
+
+          if (this.contextNode_.children && this.has(prop)) {
+            if (!newScheduled) {
+              newScheduled = scheduled.slice(0);
+            }
+            // Stop the deepscan for this value. It will be propagated
+            // by the responsible node.
+            newScheduled.push(prop.key);
+          }
+        }
+      });
+    }
+    return newScheduled || scheduled;
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  isConnected_() {
+    return !!this.contextNode_.root;
+  }
+
+  /**
+   * Start the used value tracker if it hasn't started yet.
+   *
+   * @param {!ContextProp} prop
+   * @return {!UsedDef}
+   * @private
+   */
+  startUsed_(prop) {
+    const {key, deps} = prop;
+    const usedByKey = this.usedByKey_ || (this.usedByKey_ = new Map());
+    let used = usedByKey.get(key);
+    if (!used) {
+      devAssert(deps.length <= MAX_DEPS);
+      used = {
```

Comment:
Nit: This seems like a class we should be instantiating.

Reply:
Hmm. Do you think? To highlight: there are some callbacks there, but they are not like class methods - they are arrays of callbacks for faster update. WDYT?